src/BookParser.hs

```diff
@@ -244,14 +267,22 @@ preprocessChapter = \case
     warn WInvalidChapter
     return s
 
-preprocessSection :: MonadState [Warning] n => Span -> n Span
+preprocessSection :: MonadState BookState n => Span -> n Span
 preprocessSection = \case
-  Span t -> return $ SectionRef $ SectionId t
+  Span t -> do
+     knownSections <- gets bookStateSectionsDefined
```

Comment:
Why are you doing this in `preprocessSection`, not after the chapter was parsed entirely? This way forward referencing won't work.